File: div2A/6.2.cpp
#include <iostream>
#include <algorithm>
using namespace std;

// Solve in O(N)
int main(int argc, char const *argv[])
{
	int n;
	cin >> n;
	long long a[n];
	int cum = 0;
	long long ss = 0;
	int cnt1[n];
	long long s = 0;
	if(n <3){
		cout << 0 << endl;
		return 0;
	}
	for (int i = 0; i < n; ++i)
	{
		cin >> a[i];
		s += a[i];

	}
	// forward condition
	if (s%3 != 0)
	{
		cout << 0 << endl;
		return 0;
	}
	
	// sub part size
	s /= 3;
	
	for (int i = n-1; i >= 0; --i)
	{	
		ss += a[i];
		if (ss == s)
		{
			cum += 1;
		}
		cnt1[i] = cum;
	}
	ss = 0;
	long long ways = 0;
	for (int i = 0; i < n-2; ++i)
	{	
		ss += a[i];
		if (ss == s)
		{
			ways += cnt1[i+2];
		}
	}

	cout << ways << endl;
 	return 0;
}

File: div2A/6.2.py
N = input()
A = [int(i) for i in raw_input().split()]
K = sum(A)

if K%3 != 0:
	print 0
else:
	s = sum(A)
	s /= 3
	
	# Identify third part
	cnt2 = [0 for i in range(len(A))]
	ss = 0
	cum = 0
	for i in range(len(A)-1, 1, -1):
		ss += A[i]
		if ss == s:
			cum += 1
		cnt2[i] = cum
	
	ways = 0
	# Identify first part
	cnt1 = [0 for i in range(len(A))]
	ss = 0
	for i in range(len(A)-2):
		ss += A[i]
		if ss == s:
			cnt1[i] = 1
			ways += cnt2[i+2]
	print ways